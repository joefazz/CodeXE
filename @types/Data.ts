export interface Activity {
    id: number;
    title: string;
    description: string;
    task: string;
    expectedResult: string;
    prebakedCode?: string;
    requiredCode?: string[];
}

export interface Exercise {
    id: string;
    title: string;
    slug: string;
    description: string;
    container: string;
    length: number;
    entrypoint: string;
    difficulty: 'beginner' | 'intermediate' | 'advanced';
    language: string;
    activities?: string[];
}
