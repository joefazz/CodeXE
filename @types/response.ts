import { Activity } from './Data';
import { Languages } from '@types';

export interface Exercise {
    language: Languages;
    difficulty: string;
    length: number;
    description: string;
    title: string;
    container: string;
    activities: Activity[];
    entrypoint: string;
}
