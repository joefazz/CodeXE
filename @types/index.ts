import { SetStateAction } from 'react';

import * as Response from './response';
import * as Data from './Data';

export { Data, Response };

export type Context = {
    id: string; // Container ID
    status: 'connected' | 'idle' | 'disconnected';
    socket: WebSocket;
    exerciseId: string;
    response: ContextResponse;
    containerName: string;
};

export type ContextResponse = {
    readData: any;
    writeData: any;
    metaData: any;
};

export enum Languages {
    JS = 'javascript',
    PYTHON = 'python',
    C = 'cpp'
}

export type ReactSetter<Type> = React.Dispatch<SetStateAction<Type>>;

export enum MessageTypes {
    CONTAINER_START = 'Container.Start',
    CONTAINER_EXEC = 'Container.Exec',
    CONTAINER_READ = 'Container.Read',
    CONTAINER_STOP = 'Container.Stop',
    CONTAINER_PAUSED = 'Container.Paused',
    EXERCISE_START = 'Exercise.Start',
    EXERCISE_CONNECT = 'Exercise.Connect',
    EXERCISE_STOP = 'Exercise.Stop',
    EXERCISE_RUN = 'Exercise.Run',
    EXERCISE_EXEC = 'Exercise.Exec',
    CODE_SAVE = 'Code.Save'
}
