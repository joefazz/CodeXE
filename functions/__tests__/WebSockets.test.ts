import { handleMessage } from '../WebSockets';
import { MessageTypes } from '../../@types';

const MOCK_STATE = {};

function makeEvent(type: string, data: object | string) {
    return new MessageEvent('test', { data: JSON.stringify({ type, data }) });
}

describe('Test Message Receive WebSockets', () => {
    test('Container Pause', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CONTAINER_PAUSE, {});

        expect(handleMessage(MOCK_EVENT, MOCK_STATE)).toMatchSnapshot();
    });

    test('Container Resume', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CONTAINER_RESUME, {});

        expect(handleMessage(MOCK_EVENT, MOCK_STATE)).toMatchSnapshot();
    });

    test('Container Start', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CONTAINER_START, {
            name: 'Tester',
            info: { Config: { Hostname: 'Tester' } }
        });

        expect(handleMessage(MOCK_EVENT, MOCK_STATE)).toMatchSnapshot();

        const TEMP_STATE = { containerName: '', id: '' };

        expect(handleMessage(MOCK_EVENT, TEMP_STATE)).toMatchSnapshot();
    });

    test('Container Exec', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CONTAINER_EXEC, 'Hello World');

        expect(handleMessage(MOCK_EVENT, MOCK_STATE)).toMatchSnapshot();
    });

    test('Container Read', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CONTAINER_READ, { code: '// Hello' });

        expect(handleMessage(MOCK_EVENT, MOCK_STATE)).toMatchSnapshot();
    });

    test('Stop Container', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CONTAINER_STOP, {});

        expect(handleMessage(MOCK_EVENT, MOCK_STATE)).toMatchSnapshot();
    });

    test('Connect to Exercise', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.EXERCISE_CONNECT, 'activity_id');

        expect(handleMessage(MOCK_EVENT, MOCK_STATE)).toMatchSnapshot();
    });

    test('Stop Exercise', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.EXERCISE_STOP, {});

        expect(handleMessage(MOCK_EVENT, MOCK_STATE)).toMatchSnapshot();
    });

    test('Code Saved', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CODE_SAVE, {
            success: true,
            time: 1555155949202
        });

        const TEMP_STATE = { response: { metaData: {} } };

        expect(handleMessage(MOCK_EVENT, TEMP_STATE)).toMatchSnapshot();
    });

    test('Code Read', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CODE_READ, { code: '// Hello' });

        const TEMP_STATE = { response: { readData: {} } };

        expect(handleMessage(MOCK_EVENT, TEMP_STATE)).toMatchSnapshot();
    });

    test('Get Container Filesystem', () => {
        const MOCK_EVENT = makeEvent(MessageTypes.CONTAINER_TREE, { result: ['node.js'] });

        const TEMP_STATE = { response: { metaData: {} } };

        expect(handleMessage(MOCK_EVENT, TEMP_STATE)).toMatchSnapshot();
    });
});
