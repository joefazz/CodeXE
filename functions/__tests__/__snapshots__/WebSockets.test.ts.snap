// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Message Receive WebSockets Code Read 1`] = `
Object {
  "response": Object {
    "readData": Object {
      "code": "// Hello",
    },
  },
}
`;

exports[`Test Message Receive WebSockets Code Saved 1`] = `
Object {
  "response": Object {
    "metaData": Object {
      "saveInfo": Object {
        "succeed": true,
        "timestamp": 1555155949202,
      },
    },
  },
}
`;

exports[`Test Message Receive WebSockets Connect to Exercise 1`] = `
Object {
  "activityId": "activity_id",
}
`;

exports[`Test Message Receive WebSockets Container Exec 1`] = `
Object {
  "response": Object {
    "writeData": Object {
      "output": "Hello World",
    },
  },
}
`;

exports[`Test Message Receive WebSockets Container Pause 1`] = `
Object {
  "status": "idle",
}
`;

exports[`Test Message Receive WebSockets Container Read 1`] = `
Object {
  "response": Object {
    "readData": Object {
      "code": "// Hello",
    },
  },
}
`;

exports[`Test Message Receive WebSockets Container Resume 1`] = `
Object {
  "status": "connected",
}
`;

exports[`Test Message Receive WebSockets Container Start 1`] = `
Object {
  "status": "connected",
}
`;

exports[`Test Message Receive WebSockets Container Start 2`] = `
Object {
  "containerName": "Tester",
  "id": "Tester",
  "status": "connected",
}
`;

exports[`Test Message Receive WebSockets Get Container Filesystem 1`] = `
Object {
  "response": Object {
    "metaData": Object {
      "tree": Array [
        "node.js",
      ],
    },
  },
}
`;

exports[`Test Message Receive WebSockets Stop Container 1`] = `undefined`;

exports[`Test Message Receive WebSockets Stop Exercise 1`] = `
Object {
  "activityId": "",
}
`;
